# Quantum Software Development Journey: 
# From Theory to Application with Classiq - Part 1

**Welcome to the Classiq Workshop Series for QClass 2024!**

In this series, we will develop the skills needed to participate in quantum software development!

- Week 1: Classiq's Basics & High-Level Functional Design
- Week 2: Using Git as a Tool for In-Team Collaboration and Open Source Contributions
- Weeks 3-4: Advanced Algorithms, Introduction to Quantum Machine Learning (QML), and Their Applications

**Here, you have early access to our [New Classiq's documentation](https://nightly.docs.classiq.io/latest/)!**




Additional resources you should use are
- The IDE of the classiq platform at [platform.classiq.io](platform.classiq.io)
- The [community Slack of Classiq](https://short.classiq.io/join-slack) - Classiq's team will answer any question you have over there, including implementation questions
- [Classiq's documentation](https://docs.classiq.io/latest/user-guide/platform/) with the dedicated [Python SDK explanations](https://docs.classiq.io/latest/user-guide/platform/qmod/python/functions/)

Good luck!

## Setting The Scene

Install the Classiq SDK package:

In [1]:
# !pip install -U classiq

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 6.3 MB/s eta 0:00:00a 0:00:01m
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.1/77.1 kB 12.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.6/150.6 kB 18.6 MB/s eta 0:00:00
  Using cached pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)
  Using cached tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 398.1/398.1 kB 28.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 39.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 44.0 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 92.7/92.7 kB 6.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.6/42.6 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.3/11.3 MB 31.6 MB/s eta 0:00:0000:010:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 26.2 MB/s eta 0:00:0000:0100:01
   

You need to authenticate your device in order to use Classiq's backend synthesis engine and IDE. 
**Make sure to register to the platform** at [platform.classiq.io](https://platform.classiq.io/) before you run the next cell:

In [2]:
import classiq
classiq.authenticate()

Your user code: FSWF-CZBV
If a browser doesn't automatically open, please visit this URL from any trusted device: https://auth.classiq.io/activate?user_code=FSWF-CZBV


In [3]:
from classiq import *

## A Warm Up

### First Example

Write a function that prepares the minus state $\ket{-}=\frac{1}{\sqrt2}(\ket{0}-\ket{1})$, assuming it recives the qubit $\ket{x}=\ket{0}$ (hint): 

<details>
<summary>
HINT
</summary>

Use `H(x)`,`X(x)`
</details>

In [4]:
@qfunc
def prepare_minus_state(x:QBit):
    X(x)
    hadamard_transform(x)

Now we will test our code:

In [5]:
@qfunc
def main(x: Output[QBit]):
    allocate(1,x) # Initialize the qubit x
    prepare_minus_state(x)

In [6]:
quantum_model = create_model(main)
quantum_program = synthesize(quantum_model)

In [7]:
show(quantum_program)

Opening: https://platform.classiq.io/circuit/0271d743-8117-45b4-a600-46dd5c3c7def?version=0.41.0


### Uniform Superposition

Let's continue warming up with creating a function that receives a quantum register and creates a uniform superposition for all qubits within this array. You should use the function `apply_to_all(gate_operand=, target=)`:

In [15]:
@qfunc
def create_initial_state(reg: QArray[QBit]):
    #TODO use the function apply_to_all in order create a uniform superposition
    apply_to_all(gate_operand=hadamard_transform, target=reg)

Test your function by creating a new main function, synthesizing and viewing the circuit:

In [18]:
@qfunc
def main(reg: Output[QArray[QBit]]): #TODO fill in the correct declaration here, what variables this model should output?
    #TODO allocate reg with a few qubits
    allocate(num_qubits=5, out=reg)
    create_initial_state(reg)

In [19]:
# TODO uncomment the following line:
qprog = synthesize(create_model(main))

#TODO show the quantum program
show(quantum_program)

Opening: https://platform.classiq.io/circuit/0271d743-8117-45b4-a600-46dd5c3c7def?version=0.41.0


Another implementation could utilize the `repeat(count=, iteration=)` function. The `repeat` function can be thought of as a "classical for loop". It could be handy in many situations, especially when combined with `if_` function, the "classical if" statement in Classiq. Together, they form 'Classical Control Flow'. 

Read more: [Classical Control Flow (repeat, if_)](https://docs.classiq.io/latest/user-guide/platform/qmod/language-reference/statements/classical-control-flow/?h=repea#__tabbed_1_2)

In [20]:
@qfunc
def create_initial_state(q: QArray[QBit]) -> None:
    repeat(q.len, lambda i: H(q[i]))

@qfunc
def main(reg: Output[QArray]): 
    allocate(4,reg)
    create_initial_state(reg)

qprog = synthesize(create_model(main))
show(qprog)

Opening: https://platform.classiq.io/circuit/6918f036-a65d-4a57-9f39-ee055588546b?version=0.41.0


## Guidelines for High-Level Functional Design with Classiq

**Some basic explanations about the high-level functional design with Classiq:**

* There should always be a main (`def main(...)`) function - the model that captures your algorithm is described there

* The model is always generated out of the main function 

* The model is sent to the synthesis engine (compiler) that return a quantum program which contains the quantum circuit

**Some basic guidelines about the modeling language (QMOD):**

1. Every function you use with the QMOD language should have the decorator `@qfunc` before it
2. Every quantum variable should be declared, either as an argument of a function e.g. `def prepare_minus(x: QBit)` or as a local variable within the function itself with `x = QBit('x')`


3. Some quantum variables need to be initialized with the `allocate` function. This is required in 2 cases:
* A variable is an argument of a function with the declaration `Output` like `def main(x: Output[QNum])`
* A variable that was declared within a function like `a = QNum('a')`

4. For the `main` function, you will always use `Output` for all variables. The `output` indicates that these quantum variables are not initialized outside the scope of the function.


<details> 
<summary> Types of Initializations </summary>
There are a few ways to initialize a quantum variable:

1. With `allocate` or `allocate_num` 
2. With `prepare_int`, `prepare_state` or `prepare_amplitudes`
3. As the result of a numeric operation `|=`
4. With the `bind` operation (`->` in native)
5. With any function that declares its quantum variable argument as `output`

</details>

<details> 
<summary> Types of Quantum Variables </summary>
In Qmod there are 3 types of quantum variables:

1. `QBit` (`qbit`)
2. `QArray[QBit]` (`qbit[]`)
3. `QNum` (`qnum`)

(See also [Quantum Variables](https://nightly.docs.classiq.io/latest/classiq_101/classiq_concepts/design/quantum_variables_and_functions/))
</details>

## Tutorial - State Preparation 

### Prepare State

Now, we will see how we can easily make arbitrary state using Classiq's `prepare_state`.

For example, let’s say we want to prepare the state $ \ket{\Phi^+}=\frac{1}{\sqrt{2}}(\ket{00}+\ket{11}) $

In [21]:
@qfunc
def main(x: Output[QArray[QBit]]):
    prepare_state(probabilities=[0.5,0,0,0.5], bound=0.01, out=x)

model = create_model(main)
qprog = synthesize(model)
show(qprog)

Opening: https://platform.classiq.io/circuit/8f29a5a9-1ca2-4887-acc0-dee6661510d3?version=0.41.0


Or using `prepare_bell_state` in order to prepare states with relative phase:

In [24]:
@qfunc
def main(x:Output[QArray]):
    prepare_bell_state(state_num=1, q=x) # phi-

In [25]:
quantum_model = create_model(main)
quantum_program = synthesize(quantum_model)
show(quantum_program)

Opening: https://platform.classiq.io/circuit/81f1c173-9a7d-4df3-a9eb-b5dc69091c9e?version=0.41.0


<details>
<summary>
NOTE
</summary>


| State Number | Bell State | 
|--------------|------------|
| 0            | $ \ket{\phi^+}= \frac{1}{\sqrt{2}}[\ket{00}+\ket{11}] $  | 
| 1            | $ \ket{\phi^-}= \frac{1}{\sqrt{2}}[\ket{00}-\ket{11}]$  |
| 2            | $ \ket{\psi^+}= \frac{1}{\sqrt{2}}[\ket{01}+\ket{10}]$  |
| 3            | $ \ket{\psi^-}= \frac{1}{\sqrt{2}}[\ket{01}-\ket{10}]$  |
| 4=0          | $ \ket{\phi^+}= ...$  |
| ...          | ...        | ...   | 


</details>

Now it's your turn to prepare the state $ \ket{\Psi^+}=\frac{1}{\sqrt{2}}(\ket{01}+\ket{10}) $

In [27]:
@qfunc
def main(x: Output[QArray[QBit]]):
    #TODO use the function prepare_state/ prepare_bell_state in order create a the desired state
    prepare_bell_state(state_num=2, q=x) # phi-

#TODO create model, synthesize it, and show it
model = create_model(main)
qprog = synthesize(model)
show(qprog)

Opening: https://platform.classiq.io/circuit/8d827a63-de2c-4858-8057-edaa6ab7f84f?version=0.41.0


Note that we can use `prepare_state` to create much more complex states, for instance, we can create list of probabilities:

In [28]:
probabilities = [
    0,
    0.002,
    0.004,
    0.006,
    0.0081,
    0.0101,
    0.0121,
    0.0141,
    0.0161,
    0.0181,
    0.0202,
    0.0222,
    0.0242,
    0.0262,
    0.0282,
    0.0302,
    0.0323,
    0.0343,
    0.0363,
    0.0383,
    0.0403,
    0.0423,
    0.0444,
    0.0464,
    0.0484,
    0.0504,
    0.0524,
    0.0544,
    0.0565,
    0.0585,
    0.0605,
    0.0625,
]

And then generate the state:

In [29]:
@qfunc
def main(x: Output[QArray[QBit]]):
    prepare_state(probabilities=probabilities, bound=0.01, out=x)
    
model = create_model(main)
qprog = synthesize(model)
show(qprog)

Opening: https://platform.classiq.io/circuit/812fe320-5244-4f7d-8829-d83fc7c7d777?version=0.41.0


### Prepare Int

We also have a `prepare_int` function, which allows us to register integers effortlessly. For example, $binary(9) = 1001$, or any other integer can be prepared in a single line of code:

In [30]:
from classiq import *

@qfunc
def main(x: Output[QNum]) -> None:
    prepare_int(9, x)

model = create_model(main)
qprog = synthesize(model)
show(qprog)

Opening: https://platform.classiq.io/circuit/a909dbba-7adc-4db9-88db-df392b39db1e?version=0.41.0


### Exercise: Parallel Addition 

In this exercise, we will conclude state preparation and also get a teaser of the next part of the tutorial - Arithmetic Operations. 

Let's say that for some reason we used `prepare_int` to create the integer 7, and we want to perform addition operations with the integers 0, 4, and 7. We will do that using `prepare_state`.

In [31]:
@qfunc
def main(res: Output[QNum]) -> None:
    x = QNum("x") #TODO complete the declarations of x,y
    y = QNum("y")
    
    # TODO prepare the above-mentioned integer using prepare_int
    prepare_int(7, x)
    prepare_state(probabilities=[1/3, 0,0,0, 1/3, 0 ,0, 1/3], bound=0.01, out=y) # TODO complete the 'probabilities' list
    
    res|=x+y



In [32]:
# TODO uncomment the folllowing lines:

model = create_model(main)
qprog = synthesize(model)
show(qprog)

Opening: https://platform.classiq.io/circuit/bc807842-7b3e-437a-84ef-f0aa2e69fa5f?version=0.41.0


## Tutorial - Arithmetic Operations with Classiq

One of the key advantages of Classiq is it's simplistic and powerful compiler for quantum arithmetic. Let's see an example:

In [33]:
num_qubits = 4
fraction_digits = 0 
is_signed = True

@qfunc
def main(x: Output[QNum], y: Output[QNum]):
    allocate_num(num_qubits=num_qubits, is_signed=is_signed, fraction_digits=fraction_digits, out=x)
    hadamard_transform(x)
    y|= x**2 + 1

qmod = create_model(main)

The `allocate_num` function initializes a quantum variable that represent numbers. By default, it is initialized to the $\ket{0}$ state. Then the `hadmard_transform` creates a superposition of all possible states in the domain $[-2^3,2^3-1]$. Finally, the arithmetic operation creates the entangled superposition of states:
$\begin{equation}
\sum_{x =-2^3}^{2^3-1}\ket{x}\ket{x^2+1}.
\end{equation}$

The `qmod` variable is a string that captures the algorithm we have just created in a JSON format. Now, what we want is to synthesize (compile) in order to receive a concrete quantum program that contains the quantum circuit implementation.

In [34]:
qprog = synthesize(qmod)
show(qprog)

Opening: https://platform.classiq.io/circuit/af93bf3d-a459-4bfe-b057-3502afd1c9ca?version=0.41.0


### Advanced Arithmetics

Now let's create a general linear function with Classiq: $y= ax+b$ where $a,b$ are classical integer parameters and $x,y$ is a quantum states representing integers:

In [61]:
@qfunc
def linear_func(a:CInt,b: CInt, x:QNum, y: Output[QNum]):
    y |= a*x+b

In [36]:
@qfunc
def main(x:Output[QNum], y: Output[QNum]):

    a = 2
    b = 1
    allocate_num(num_qubits=4,is_signed=False,fraction_digits=0,out=x)
    hadamard_transform(x)
    linear_func(a,b,x,y)

qmod = create_model(main)

In [37]:
qprog = synthesize(qmod)

Let's execute the circuit from directly from the SDK:

In [38]:
job = execute(qprog)

And we can view the results in the IDE:

In [39]:
job.open_in_ide()

Or to directly analyze it within the SDK:

In [40]:
results = job.result()
parsed_counts = results[0].value.parsed_counts
for sampled_state in parsed_counts: print(sampled_state.state)

{'x': 2.0, 'y': 5.0}
{'x': 6.0, 'y': 13.0}
{'x': 3.0, 'y': 7.0}
{'x': 0.0, 'y': 1.0}
{'x': 8.0, 'y': 17.0}
{'x': 5.0, 'y': 11.0}
{'x': 7.0, 'y': 15.0}
{'x': 12.0, 'y': 25.0}
{'x': 1.0, 'y': 3.0}
{'x': 11.0, 'y': 23.0}
{'x': 13.0, 'y': 27.0}
{'x': 14.0, 'y': 29.0}
{'x': 4.0, 'y': 9.0}
{'x': 15.0, 'y': 31.0}
{'x': 10.0, 'y': 21.0}
{'x': 9.0, 'y': 19.0}


**Now it's your turn!** 

Implement the same linear function, but now $x$ is in the domain $[0,1)$ and is represented by 4 qubits. The parameters $a,b$ should be now `float` with the values of: $a=0.5, b=1.5$.

In [60]:
@qfunc
def linear_func2(a:CReal,b: CReal, x:QNum, y: Output[QNum]):
    y |= a*x+b
    
@qfunc
def main(x:Output[QNum], y: Output[QNum]):

    a = 0.5
    b = 1.5
    allocate_num(num_qubits=4,is_signed=False,fraction_digits=4,out=x)
    hadamard_transform(x)
    linear_func2(a,b,x,y)

qmod = create_model(main)
qprog = synthesize(qmod)
job = execute(qprog)

In [53]:
results = job.result()
parsed_counts = results[0].value.parsed_counts
for sampled_state in parsed_counts: print(sampled_state.state)

{'x': 0.0625, 'y': 1.53125}
{'x': 0.4375, 'y': 1.71875}
{'x': 0.1875, 'y': 1.59375}
{'x': 0.25, 'y': 1.625}
{'x': 0.375, 'y': 1.6875}
{'x': 0.0, 'y': 1.5}
{'x': 0.8125, 'y': 1.90625}
{'x': 0.6875, 'y': 1.84375}
{'x': 0.5, 'y': 1.75}
{'x': 0.75, 'y': 1.875}
{'x': 0.875, 'y': 1.9375}
{'x': 0.9375, 'y': 1.96875}
{'x': 0.125, 'y': 1.5625}
{'x': 0.625, 'y': 1.8125}
{'x': 0.3125, 'y': 1.65625}
{'x': 0.5625, 'y': 1.78125}


### Tutorial - Two controlled Linear operations

Let's say we want now to have two linear operations applied on the same quantum variable (register). But the arithmetic operation initialize a new quantum variable, so how can we do that? The answer is that we need to apply the operation to another variable and then XOR it to the variable we want. 

This can be useful if the linear operation we want to apply is controlled upon a control variable. Let's first define the functional building block:

In [62]:
@qfunc
def inplace_linear_attempt(a:CInt,b: CInt, x:QNum, y: QNum):
    tmp = QNum('tmp')
    linear_func(a,b,x,tmp)
    inplace_xor(tmp,y)

And checking our basic function implementation:

In [63]:
@qfunc
def main(x: Output[QNum],y: Output[QNum]):
    a = 1
    b = 2

    allocate_num(4,False,0,y)
    allocate_num(4,False,0,x)
    hadamard_transform(x)
    inplace_linear_attempt(a,b,x,y)

qmod = create_model(main)
qprog = synthesize(qmod)
show(qprog)

Opening: https://platform.classiq.io/circuit/f09b4bff-fba8-4d90-83a8-f4acf6543af6?version=0.41.0


OK, cool. So now we want to add a control qubit that controlled on the state $\ket{0}$ implements the linear function $\ket{x}\rightarrow\ket{x}\ket{x+2}$ and controlled on the state $\ket{1}$ implements the linear function $\ket{x}\rightarrow\ket{x}\ket{2x+1}$:

In [64]:
@qfunc
def control_logic(a: CArray[CInt], b: CArray[CInt], controller:QNum, x: QNum, y: QNum):
    
    repeat( count=a.len,         
            iteration=lambda i: control(controller==i, lambda: inplace_linear_attempt(a[i],b[i],x,y)))


In [65]:
@qfunc
def main(controller: Output[QNum], x: Output[QNum],y: Output[QNum]):

    # Linear polynom parameters
    a = [1,2]
    b = [2,1]

    # Initializing x to a superposition in the domain [0,2^4-1]
    allocate_num(4,False,0,x)
    hadamard_transform(x)
    
    #Initialize y
    allocate_num(4,False,0,y)

    # Setting the controller in a superposition
    allocate_num(1,False,0,controller)
    H(controller)

    # Implementing the control logic
    control_logic(a,b,controller,x,y)

    
qmod = create_model(main)

In [66]:
qprog = synthesize(qmod)
show(qprog)

Opening: https://platform.classiq.io/circuit/5ddd48f7-190b-426c-b8fd-9bde42324a91?version=0.41.0


By executing we can actually see we get what we want:

In [67]:
def print_parsed_counts(job):
    results = job.result()
    parsed_counts = results[0].value.parsed_counts
    for parsed_state in parsed_counts: print(parsed_state.state)

job = execute(qprog)
print_parsed_counts(job)

{'controller': 1.0, 'x': 5.0, 'y': 11.0}
{'controller': 0.0, 'x': 2.0, 'y': 4.0}
{'controller': 0.0, 'x': 13.0, 'y': 15.0}
{'controller': 0.0, 'x': 11.0, 'y': 13.0}
{'controller': 0.0, 'x': 6.0, 'y': 8.0}
{'controller': 1.0, 'x': 7.0, 'y': 15.0}
{'controller': 0.0, 'x': 12.0, 'y': 14.0}
{'controller': 1.0, 'x': 10.0, 'y': 5.0}
{'controller': 0.0, 'x': 1.0, 'y': 3.0}
{'controller': 1.0, 'x': 12.0, 'y': 9.0}
{'controller': 0.0, 'x': 15.0, 'y': 1.0}
{'controller': 1.0, 'x': 2.0, 'y': 5.0}
{'controller': 0.0, 'x': 0.0, 'y': 2.0}
{'controller': 1.0, 'x': 8.0, 'y': 1.0}
{'controller': 1.0, 'x': 11.0, 'y': 7.0}
{'controller': 1.0, 'x': 4.0, 'y': 9.0}
{'controller': 0.0, 'x': 3.0, 'y': 5.0}
{'controller': 1.0, 'x': 0.0, 'y': 1.0}
{'controller': 1.0, 'x': 15.0, 'y': 15.0}
{'controller': 0.0, 'x': 8.0, 'y': 10.0}
{'controller': 1.0, 'x': 3.0, 'y': 7.0}
{'controller': 0.0, 'x': 9.0, 'y': 11.0}
{'controller': 1.0, 'x': 13.0, 'y': 11.0}
{'controller': 0.0, 'x': 10.0, 'y': 12.0}
{'controller': 1.0, 

Of course there is the issue of rounding and overflow - when one tries to represent $2*15+1=31$ with $4$ binary digits that's not possible (because the domain $[0,31]$ of integers is represented by at least 5 bits). See our [documentation](https://docs.classiq.io/latest/user-guide/platform/qmod/python/quantum-expressions/#inplace-arithmetic-operators) for further explanations.

Let's try to use Classiq and optimize the circuit for minimal circuit width:

In [68]:
def print_depth_width(quantum_program):
    generated_circuit = QuantumProgram.parse_raw(quantum_program)
    print(f"Synthesized circuit width: {generated_circuit.data.width}, depth: {generated_circuit.transpiled_circuit.depth}")
 
qmod = set_constraints(qmod,Constraints(optimization_parameter='width'))
qprog = synthesize(qmod)
print_depth_width(qprog)

Synthesized circuit width: 20, depth: 688


And when optimizing for depth:

In [69]:
qmod = set_constraints(qmod,Constraints(optimization_parameter='depth'))
qprog = synthesize(qmod)
print_depth_width(qprog)

Synthesized circuit width: 32, depth: 398


**Firstly, we can see here a clear demonstration of the power of high-level functional design!** The same algorithm with the same functionality was optimized once for depth and once for width and the result is 2 different circuits with different characteristics that implement the same functionality.

Secondly, is this the best we can do? Obviously the Classiq synthesis engine is optimizing for us the algorithm quite good. But, can we change something with our functionality, with our algorithm to get more efficient circuits? 

If we go back to out `inplace_linear_attempt` function, we can see that we initialize a `tmp` variable that requires more qubits and is not used. For such scenarios we have the `within_apply`. This logic implements sort of $UVU^{\dagger}$ and when temporary variables are outputs of $U$ and used only by $V$ they are uncomputed by $U^{\dagger}$. Let's see for our example:

In [70]:
@qfunc
def inplace_linear_func(a:CInt,b: CInt, x:QNum, y: QNum):
    tmp = QNum('tmp')
    within_apply(compute= lambda: linear_func(a,b,x,tmp),
                action= lambda: inplace_xor(tmp,y))

With the new `control_logic`:

In [71]:
@qfunc
def control_logic_2(a: CArray[CInt], b: CArray[CInt], controller:QNum, x: QNum, y: QNum):
    
    repeat( count=a.len,         
            iteration=lambda i: control(controller==i, lambda: inplace_linear_func(a[i],b[i],x,y)))

And when we put all together now:

In [72]:
@qfunc
def main(controller: Output[QNum], x: Output[QNum],y: Output[QNum]):

    # Linear polynom parameters
    a = [1,2]
    b = [2,1]

    # Initializing x to a superposition in the domain [0,2^4-1]
    allocate_num(4,False,0,x)
    hadamard_transform(x)
    
    #Initialize y
    allocate_num(4,False,0,y)

    # Setting the controller in a superposition
    allocate_num(1,False,0,controller)
    H(controller)

    # Implementing the control logic
    control_logic_2(a,b,controller,x,y)

    
qmod = create_model(main)

In [73]:
qprog = synthesize(qmod)

In [74]:
show(qprog)

Opening: https://platform.classiq.io/circuit/1b37f533-7e54-4085-8efa-3fb61cca0a0d?version=0.41.0


And now when we optimize for width:

In [75]:
qmod = set_constraints(qmod,Constraints(optimization_parameter='width'))
qprog = synthesize(qmod)
print_depth_width(qprog)

Synthesized circuit width: 16, depth: 244


And for depth:

In [76]:
qmod = set_constraints(qmod,Constraints(optimization_parameter='depth'))
qprog = synthesize(qmod)
print_depth_width(qprog)

Synthesized circuit width: 45, depth: 203


So using the `within_apply` logic enabled us to reduce the optimal circuit implementation in terms of width from $20$ to $16$ and the optimal circuit depth from $398$ to $203$! 

## Cheat Sheet

### Initalizations

In [77]:
allocate(
    num_qubits: CInt,
    out: Output[QArray[QBit, Literal["num_qubits"]]])
    '''
    x = QArray('x')
    allocate(4,x)
    '''

allocate_num(
    num_qubits: CInt,
    is_signed: QParam[bool],
    fraction_digits: CInt,
    out: Output[QNum])
'''
x = QNum('x')
allocate_num(4,False,4,x)
'''

SyntaxError: invalid syntax (519441306.py, line 2)

### Operations

In [ ]:
repeat(
    count: CInt,
    iteration: QCallable[CInt],
)
'''
x = QArray
allocate(4,x)
repeat(x.len,lambda index: H(x))
'''

control(
    operand: QCallable,
    ctrl: QArray[QBit],
) 
'''
x = QArray('x')
y = QArray('y')
x = allocate(4,x)
y = allocate(4,y)
repeat(x.len,lambda i: control(lambda: X(y[i]),x[i]))
'''